src/validation.cpp

```diff
@@ -1061,28 +1178,131 @@ MempoolAcceptResult MemPoolAccept::AcceptSingleTransaction(const CTransactionRef
+        return MempoolAcceptResult::Success(std::move(ws.m_replaced_transactions), ws.m_base_fees);
     }
 
-    if (!Finalize(args, ws)) return MempoolAcceptResult(ws.m_state);
+    if (!Finalize(args, ws)) return MempoolAcceptResult::Failure(ws.m_state);
 
     GetMainSignals().TransactionAddedToMempool(ptx, m_pool.GetAndIncrementSequence());
 
-    return MempoolAcceptResult(std::move(ws.m_replaced_transactions), ws.m_base_fees);
+    return MempoolAcceptResult::Success(std::move(ws.m_replaced_transactions), ws.m_base_fees);
+}
+
+PackageMempoolAcceptResult MemPoolAccept::AcceptMultipleTransactions(const std::vector<CTransactionRef>& txns, ATMPArgs& args)
+{
+    AssertLockHeld(cs_main);
+
+    PackageValidationState package_state;
+    const unsigned int package_count = txns.size();
+
+    // These context-free package limits can be checked before taking the mempool lock.
+    if (package_count > MAX_PACKAGE_COUNT) {
+        package_state.Invalid(PackageValidationResult::PCKG_POLICY, "too-many-transactions");
+        return PackageMempoolAcceptResult(package_state, {});
+    }
+
+    const int64_t total_size = std::accumulate(txns.cbegin(), txns.cend(), 0,
+                               [](int64_t sum, const auto& tx) { return sum + GetVirtualTransactionSize(*tx); });
+    if (total_size > MAX_PACKAGE_SIZE * 1000) {
+        package_state.Invalid(PackageValidationResult::PCKG_POLICY, "too-large");
+        return PackageMempoolAcceptResult(package_state, {});
+    }
+
+    std::vector<Workspace> workspaces{};
+    workspaces.reserve(package_count);
+
+    // Require the package to be sorted in order of dependency, i.e. parents appear before children.
+    // An unsorted package will fail anyway on missing-inputs, but it's better to quit earlier and
+    // fail on something less ambiguous (missing-inputs could also be an orphan or trying to
+    // spend nonexistent coins).
+    for (const auto& tx : txns) {
+        Workspace workspace(tx);
+        for (const auto& input : tx->vin) {
+            if (std::find_if(workspaces.begin(), workspaces.end(),
+                             [&, txid = input.prevout.hash](const auto& preceding_ws)
+                             { return preceding_ws.m_ptx->GetHash() == txid; }) == workspaces.end() &&
+                std::find_if(txns.begin(), txns.end(),
+                             [&, txid = input.prevout.hash](const auto& tx)
+                             { return tx->GetHash() == txid; }) != txns.end()) {
+                // If we don't find the parent in workspaces but we do find it in txns,
+                // then it must be a subsequent transaction in the package.
+                package_state.Invalid(PackageValidationResult::PCKG_POLICY, "package-not-sorted");
+                return PackageMempoolAcceptResult(package_state, {});
+            }
+        }
+        workspaces.emplace_back(std::move(workspace));
```

Comment:
nit: you can directly construct the workspace on the back of workspaces

Reply:
altho then you want to go to end-1 if you do that.

Reply:
what's end-1?

Reply:
constructing in place done